datatables.qmd

```diff
@@ -0,0 +1,255 @@
+
+Many external representations of data tables, such as in comma-separated-value (CSV) files, are row-oriented.
+Furthermore, it is often convenient to generate a data table a row at a time.
+
+Thus it becomes convenient to have a "clearing house" that can accept either row tables or column tables and provided the desired form to a downstream package, such as [StatsModels.jl](https://github.com/JuliaStats/StatsModels.jl).
+[Tables.jl](https://github.com/JuliaData/Tables.jl) does exactly this.
+It is not an implementation of data tables itself, but rather it defines "an interface for tables in Julia", allowing a row-oriented table to be accessed column-wise and vice-versa.
+
+It defines a prototype column-oriented table, `Tables.ColumnTable`, as a [NamedTuple](https://docs.julialang.org/en/v1/base/base/#Core.NamedTuple) of vectors.
+
+```{julia}
+Tables.ColumnTable
+```
+
+and a prototype row-oriented table as a vector of `NamedTuple`s.
+
+```{julia}
+Tables.RowTable
+```
+
+The actual implementation of a row-table or column-table type may be different from these prototypes but it must provide access methods as if it were one of these types.
+`Tables.jl` provides the "glue" to treat a particular data table type as if it were row-oriented, by calling `Tables.rows` or `Tables.rowtable` on it, or column-oriented, by calling `Tables.columntable` on it.
+
+## The Table type from TypedTables
+
+[TypedTables.jl](https://github.com/JuliaData/TypedTables.jl) is a lightweight package (about 1500 lines of source code) that provides a concrete implementation of column-tables, called simply `Table`, as a `NamedTuple` of vectors.
+
+A `Table` that is constructed from another type of column-table, such as an `Arrow.Table` or a `DataFrame` or an explicit `NamedTuple` of vectors, is simply a wrapper around the original table's contents.
+On the other hand, constructing a `Table` from a row table first creates a `ColumnTable`, then wraps it.
+
+```{julia}
+contratbl = Table(contra)
+```
+
+```{julia}
+typeof(contratbl)
+```
+
+(The output from that expression is a very long string.  You need to scroll to the right over the output to see all the output.)
+
+This type of table is said to be "strongly typed", meaning that the data type itself contains a wealth of detail about the exact form of the table, allowing the Julia compiler to generate efficient code for operations on the table.
+That is the positive aspect of being so specific about the names of the columns and the details of the type of data in each column.
+However, it also means that this mechanism is not suitable for tables with a large number, say hundreds or thousands, of columns, which can overburden the compiler.
+
+### Accessing columns or rows in a Table
+
+The methods for accessing columns or rows in a `Table` are simple.
+
+A column is accessed by its name as a "property", either using the `getproperty` extractor function or, more commonly, with the dot (`.`) operator, returning a vector.
```

Comment:
(thinking out loud)

this is true for `Table`, `ColumnTable` and `DataFrame`, but I don't know if this is true for all column tables. `Tables.getcolumn` is a canonical method for this type of access but it's probably too much to discuss that here.

Reply:
Just making a general comment. The section header does indicate that you're focusing on `TypedTables.Table`. :slightly_smiling_face: